DOCSP-43085: Compound Indexes (#46)
mongodb/docs-c: source/indexes/compound-index.txt

.. _c-compound-index:

================
Compound Indexes
================

.. contents:: On this page
   :local:
   :backlinks: none
   :depth: 2
   :class: singlecol

.. facet::
   :name: genre
   :values: reference
 
.. meta::
   :keywords: index, query, optimization, efficiency

Overview
--------

:manual:`Compound indexes </core/index-compound/>` hold references to multiple

> fields within a collection's documents, improving query and sort performance.

Reviewer: S: add a sentence like "Use the ``mongoc_collection_create_indexes_with_opts()`` method to create an index.

When creating a compound index, you must specify the following:

> - The fields on which to create the index. 

Reviewer: ```suggestion
- The fields to index. 
```

- The sort order for each field (ascending or descending). Specify ``BCON_INT32 (1)``
  for ascending order and ``BCON_INT32 (-1)`` for descending order.

Sample Data
~~~~~~~~~~~

The examples in this guide use the ``movies`` collection in the ``sample_mflix``
database from the :atlas:`Atlas sample datasets </sample-data>`. To learn how to create a
free MongoDB Atlas cluster and load the sample datasets, see the
:atlas:`Get Started with Atlas </getting-started>` guide.

Create a Compound Index
-----------------------

The following example creates a compound index on the ``type`` and ``genres`` fields, with
both fields indexed in ascending order:

.. literalinclude:: /includes/indexes/indexes.c
   :start-after: start-index-compound
   :end-before: end-index-compound
   :language: c
   :copyable:
   :dedent:

> The following is an example of a query that uses the index created in
> the preceding code sample:

Reviewer: ```suggestion
The following example performs a query that uses the index created in
the preceding code:
```
